### PR TITLE
Shopping cart

### PR DESCRIPTION
**Pull request by ihorromaniuk**

### CODE REVIEW

File: src/main/java/core/basesyntax/bookstore/controller/CartController.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                    description = "Didn't find the book by id",
+                    content = @Content(schema = @Schema(implementation = ExceptionDto.class))
+            )
+    })
+    @PreAuthorize("hasRole('USER')")
+    @PostMapping
+    public CartItemWithoutBookDto addItemToShoppingCart(@RequestBody
+                                                        @Valid
+                                                        AddCardItemRequestDto requestDto,
+                                                        Authentication authentication) {
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/bookstore/controller/CartController.java]:
```suggestion
    public CartItemWithoutBookDto addItemToShoppingCart(
        @RequestBody @Valid AddCardItemRequestDto requestDto,
        Authentication authentication
    ) {
```
I recommend using this formatting. As for me, it’s  more readable

File: src/main/java/core/basesyntax/bookstore/dto/cart/AddCardItemRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package core.basesyntax.bookstore.dto.cart;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+
+public record AddCardItemRequestDto(@NotNull Long bookId,
+                                    @Min(1) int quantity) {
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/bookstore/dto/cart/AddCardItemRequestDto.java]: better use Positive

File: src/main/java/core/basesyntax/bookstore/dto/cart/CartItemWithoutBookDto.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.bookstore.dto.cart;
+
+public record CartItemWithoutBookDto(Long bookId,
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/bookstore/dto/cart/CartItemWithoutBookDto.java]: without book?

File: src/main/java/core/basesyntax/bookstore/dto/cart/UpdateCartItemQuantityRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package core.basesyntax.bookstore.dto.cart;
+
+import jakarta.validation.constraints.Min;
+
+public record UpdateCartItemQuantityRequestDto(@Min(1) int quantity) {
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/bookstore/dto/cart/UpdateCartItemQuantityRequestDto.java]: the same

File: src/main/java/core/basesyntax/bookstore/mapper/CartMapper.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Mapping(target = "bookId", source = "book.id")
+    CartItemWithoutBookDto toCartItemWithoutDto(CartItem cartItem);
+
+    @Mapping(target = "userId", source = "user.id")
+    @Mapping(target = "cartItems", source = "cartItems", qualifiedByName = "setShoppingCarts")
+    ShoppingCartDto toShoppingCartDto(ShoppingCart shoppingCart);
+
+    @Named("setShoppingCarts")
+    default List<CartItemDto> setShoppingCarts(Set<CartItem> cartItems) {
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/bookstore/mapper/CartMapper.java]: Maybe setCartItems?

File: src/main/java/core/basesyntax/bookstore/model/ShoppingCart.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+@Getter
+@Setter
+@Entity
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @OneToOne(fetch = FetchType.LAZY)
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/bookstore/model/ShoppingCart.java:31]: better use MapsId

File: src/main/java/core/basesyntax/bookstore/model/ShoppingCart.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.HashSet;
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "shopping_carts")
+public class ShoppingCart {
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/bookstore/model/ShoppingCart.java:26]: should it be a soft-deleted entity?

File: src/main/java/core/basesyntax/bookstore/service/impl/CartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Override
+    public void deleteCartItem(Long cartItemId, User user) {
+        ShoppingCart shoppingCart = shoppingCartRepository.getByUser(user);
+        if (cartItemRepository.existsByIdAndShoppingCart(cartItemId, shoppingCart)) {
+            cartItemRepository.deleteById(cartItemId);
+        } else {
+            throw new EntityNotFoundException("Can't find cart item by id. Id: " + cartItemId);
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/bookstore/service/impl/CartServiceImpl.java]: better avoid if-else construction

File: src/main/java/core/basesyntax/bookstore/service/impl/UserServiceImpl.java [hunk 9]
```diff
@@ -36,6 +39,12 @@ public UserDto save(UserRegistrationRequestDto requestDto) {
                 new EntityNotFoundException("Can't find role by name. "
                         + "Name: " + Role.RoleName.USER));
         user.setRoles(Set.of(userRole));
-        return userMapper.toDto(userRepository.save(user));
+        userRepository.save(user);
+
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/bookstore/service/impl/UserServiceImpl.java]: should you use the Transactional above method

File: src/main/resources/db/changelog/changes/08-create-shopping_carts-table.yaml [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+              - column:
+                  name: user_id
+                  type: BIGINT
+                  constraints:
+                    nullable: false
+            tableName: shopping_carts
+  - changeSet:
+      id: add-unique-constraint-user_id
+      author: Ihor Romaniuk
+      objectQuotingStrategy: QUOTE_ONLY_RESERVED_WORDS
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/08-create-shopping_carts-table.yaml]: why do you need it?

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/core/basesyntax/bookstore/service/impl/CartServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    public CartItemWithBookTitleDto updateCartQuantity(Long cartItemId,
+                                                       UpdateCartItemQuantityRequestDto requestDto,
+                                                       User user) {
+        ShoppingCart shoppingCart = shoppingCartRepository.getByUser(user);
+        CartItem cartItem = cartItemRepository.findByIdAndShoppingCart(cartItemId, shoppingCart)
+                .orElseThrow(() ->
+                        new EntityNotFoundException("Can't find cart item by id. "
+                                + "Id: " + cartItemId));
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/bookstore/service/impl/CartServiceImpl.java]: did you provide enough details here?

File: src/main/java/core/basesyntax/bookstore/service/impl/CartServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Override
+    public void deleteCartItem(Long cartItemId, User user) {
+        ShoppingCart shoppingCart = shoppingCartRepository.getByUser(user);
+        if (cartItemRepository.existsByIdAndShoppingCart(cartItemId, shoppingCart)) {
+            cartItemRepository.deleteById(cartItemId);
+        }
+
+        throw new EntityNotFoundException("Can't find cart item by id. Id: " + cartItemId);
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/bookstore/service/impl/CartServiceImpl.java]: this error is always thrown. Is this a correct behavior?

**Review by sarakhmen — CHANGES_REQUESTED** [review]
let's improve your solution :)

File: src/main/java/core/basesyntax/bookstore/service/impl/CartServiceImpl.java (hunk 11 above)
Comment (ihorromaniuk) [reply src/main/java/core/basesyntax/bookstore/service/impl/CartServiceImpl.java]: yes, user and therefore shopping cart comes from authentication

File: src/main/java/core/basesyntax/bookstore/controller/CartController.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+            ),
+            @ApiResponse(
+                    responseCode = "404",
+                    description = "Didn't find the book by id",
+                    content = @Content(schema = @Schema(implementation = ExceptionDto.class))
+            )
+    })
+    @PreAuthorize("hasRole('USER')")
+    @PostMapping
+    public CartItemDto addItemToShoppingCart(
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/controller/CartController.java]:
```suggestion
    public ShoppingCartDto addItemToShoppingCart(
```

File: src/main/java/core/basesyntax/bookstore/controller/CartController.java [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+            ),
+            @ApiResponse(
+                    responseCode = "404",
+                    description = "Didn't find cart item by id",
+                    content = @Content(schema = @Schema(implementation = ExceptionDto.class))
+            )
+    })
+    @PreAuthorize("hasRole('USER')")
+    @PutMapping("/items/{cartItemId}")
+    public CartItemWithBookTitleDto updateCartItemQuantity(
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/controller/CartController.java]:
```suggestion
    public ShoppingCartDto updateCartItemQuantity(
```

File: src/main/java/core/basesyntax/bookstore/dto/cart/AddCardItemRequestDto.java [hunk 15]
```diff
@@ -0,0 +1,8 @@
+package core.basesyntax.bookstore.dto.cart;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+
+public record AddCardItemRequestDto(@NotNull Long bookId,
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/dto/cart/AddCardItemRequestDto.java]:
```suggestion
public record AddCardItemRequestDto(@NotNull @Positive Long bookId,
```

File: src/main/java/core/basesyntax/bookstore/model/ShoppingCart.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id=?")
+@SQLRestriction("is_deleted = false")
+@Entity
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    private Long id;
+
+    @MapsId
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/model/ShoppingCart.java]:
```suggestion
    @OneToOne(fetch = FetchType.LAZY)
```

File: src/main/java/core/basesyntax/bookstore/model/ShoppingCart.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+    @JoinColumn(name = "id")
+    private User user;
+
+    @OneToMany(mappedBy = "shoppingCart")
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/model/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/core/basesyntax/bookstore/model/User.java [hunk 18]
```diff
@@ -22,6 +23,7 @@
 @Setter
 @SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id=?")
 @SQLRestriction("is_deleted=false")
+@NoArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/model/User.java]: Remove
```suggestion
```

File: src/main/java/core/basesyntax/bookstore/service/impl/CartServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+import core.basesyntax.bookstore.repository.cart.CartItemRepository;
+import core.basesyntax.bookstore.repository.cart.ShoppingCartRepository;
+import core.basesyntax.bookstore.service.BookService;
+import core.basesyntax.bookstore.service.CartService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CartServiceImpl implements CartService {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/service/impl/CartServiceImpl.java]: all logic should be in ShoppingCartService, remove CartServiceImpl

File: src/main/java/core/basesyntax/bookstore/service/impl/UserServiceImpl.java [hunk 20]
```diff
@@ -36,6 +41,12 @@ public UserDto save(UserRegistrationRequestDto requestDto) {
                 new EntityNotFoundException("Can't find role by name. "
                         + "Name: " + Role.RoleName.USER));
         user.setRoles(Set.of(userRole));
-        return userMapper.toDto(userRepository.save(user));
+        userRepository.save(user);
+
+        ShoppingCart shoppingCart = new ShoppingCart();
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/service/impl/UserServiceImpl.java]: create method in ShoppingCartService `createShoppingCart(User user)` to create shop cart for user

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/core/basesyntax/bookstore/model/ShoppingCart.java (hunk 17 above)
Comment (ihorromaniuk) [reply src/main/java/core/basesyntax/bookstore/model/ShoppingCart.java:33]: already changed in order task

**Review by sarakhmen — APPROVED** [review]
lgtm!

**Pull request merged**